lib/logstash/inputs/syslog.rb

```diff
@@ -135,37 +155,43 @@ def udp_listener(output_queue)
   def tcp_listener(output_queue)
     @logger.info("Starting syslog tcp listener", :address => "#{@host}:#{@port}")
     @tcp = TCPServer.new(@host, @port)
-    @tcp_clients = []
 
     loop do
-      client = @tcp.accept
-      @tcp_clients << client
-      Thread.new(client) do |client|
-        ip, port = client.peeraddr[3], client.peeraddr[1]
-        @logger.info("new connection", :client => "#{ip}:#{port}")
-        LogStash::Util::set_thread_name("input|syslog|tcp|#{ip}:#{port}}")
-        begin
-          client.each do |line|
-            @codec.decode(line) do |event|
-              decorate(event)
-              event["host"] = ip
-              syslog_relay(event)
-              output_queue << event
-            end
-          end
-        rescue Errno::ECONNRESET
-        ensure
-          @tcp_clients.delete(client)
-        end
-      end # Thread.new
-    end # loop do
+      socket = @tcp.accept
+      @tcp_sockets << socket
+
+      break if @shutdown_requested.true?
+
+      Thread.new(output_queue, socket) do |output_queue, socket|
+        tcp_receiver(output_queue, socket)
+      end
+    end
   ensure
     close_tcp
   end # def tcp_listener
 
+  def tcp_receiver(output_queue, socket)
+    ip, port = socket.peeraddr[3], socket.peeraddr[1]
+    @logger.info("new connection", :client => "#{ip}:#{port}")
+
+    LogStash::Util::set_thread_name("input|syslog|tcp|#{ip}:#{port}}")
+
+    socket.each do |line|
+      @codec.decode(line) do |event|
+        decorate(event)
+        event["host"] = ip
+        syslog_relay(event)
+        output_queue << event
+      end
```

Comment:
`@codec.decode()` in the tcp receiver and the udp receiver implement the same logic, the only different part is the source of the IP address we could refactor this into one method.